src/coreclr/jit/lclmorph.cpp

```diff
@@ -2426,3 +2426,562 @@ void Compiler::fgExposeLocalsInBitVec(BitVec_ValArg_T bitVec)
+    }
+    else
+    {
+        // Convert the input of the store to a mask.
+        assert(weight.simdBaseJitType != CORINFO_TYPE_UNDEF);
+        GenTree* convertOp =
+            gtNewSimdCvtVectorToMaskNode(TYP_MASK, lclStore->Data(), weight.simdBaseJitType, weight.simdSize);
+        lclStore->Data() = convertOp;
+
+        JITDUMP("Updated V%02d to store as mask (Added conversion)\n", lclStore->GetLclNum());
+    }
+
+#ifdef DEBUG
+    if (verbose)
+    {
+        gtDispTree(lclStore);
+    }
+#endif
+
+    return true;
+}
+
+//-----------------------------------------------------------------------------
+// fgLCLMasksUpdateLCLVar: For the given lcl var, if the weighting recommends to switch,
+// then update to use the source as a mask.
+//
+// Arguments:
+//     lclVar - The local variable.
+//     stmt - The statement the local vairable is contained in.
+//     weightsTable - table to update.
+//
+void Compiler::fgLCLMasksUpdateLCLVar(GenTreeLclVarCommon* lclVar,
+                                      Statement* const     stmt,
+                                      LCLMasksWeightTable* weightsTable)
+{
+    if (!lclVar->OperIs(GT_LCL_VAR))
+    {
+        return;
+    }
+
+    LCLMasksWeight weight;
+    bool           found = weightsTable->Lookup(lclVar->GetLclNum(), &weight);
+
+    // If there no entry, then the var does not have a local store.
+    if (!found)
+    {
+        return;
+    }
+
+    if (!weight.ShouldSwitch())
+    {
+        JITDUMP("Local Var V%02d at [%06u] will not be converted. ", lclVar->GetLclNum(), dspTreeID(lclVar));
+        weight.DumpTotalWeight();
+        return;
+    }
+
+    JITDUMP("Local Var V%02d at [%06u] will be converted. ", lclVar->GetLclNum(), dspTreeID(lclVar));
+    weight.DumpTotalWeight();
+
+    // Remove or add a mask conversion/
+    LCLMasksUpdateLCLVarVisitor ev(this, lclVar->GetLclNum(), stmt, weight.simdBaseJitType, weight.simdSize);
+    GenTree*                    root = stmt->GetRootNode();
+    ev.WalkTree(&root, nullptr);
+}
+
+#endif // TARGET_ARM64
+
+//------------------------------------------------------------------------
+// optLCLMasks: Allow locals to be of Mask type
+//
+// At the C# level, Masks share the same type as a Vector. It's possible for the same
+// variable to be used as a mask or vector. Any APIs that return a mask must first convert
+// the value to a vector before storing it to a variable. Any uses of a variable as a mask
+// must first convert from vector before using it. In many cases this creates unnecessary
+// conversions. For variables that live outside the scope of the current method then the
+// conversions are required to ensure correctness. However, for local variables where the
+// scope is local to the current method, then it is possible to keep the value as a mask,
+// by updating all definitions and uses.
+//
+// In the common case it is expected that uses of masks are consistent - once a variable is
```

Comment:
thanks for writing this up.

Reply:
I would also include little portion of hash table implementation and how we decide what is profitable to eliminate vs. add conversion for.